Add resource_from_fqcr utility function

felixfontein opened the pull request:
##### SUMMARY
~~Currently become plugins in community.general are broken (if `become_exe` is not explicitly specified by the user), since~~ https://github.com/ansible/ansible/blob/01638e0ea222b115909bfd73677b97be92744ff2/lib/ansible/playbook/play_context.py#L358 ~~explicitly uses the `become_name` as a default value, which for become plugins in community.general is prefixed with `community.general.`.~~ (No longer needed.)

Also adds a helper method `resource_from_fqcr` to collection_loader, which avoids the tedious `.split('.')[-1]` found in several places.

##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME
```
lib/ansible/executor/powershell/module_manifest.py
lib/ansible/playbook/play_context.py
lib/ansible/plugins/action/__init__.py
lib/ansible/plugins/cache/__init__.py
lib/ansible/utils/collection_loader.py
```

felixfontein committed:
Add resource_from_fqcr utility function.

bcoca reviewed (changes requested):
Instead of enabling this,  i would just require that all collection plugins set a default and even would deprecate the 'name' fallback for existing 3rd party plugins that currently use this outside of collections.

felixfontein commented:
That would also be something I would prefer. I'm not sure how to deprecate it properly, though. Right now all plugins I know of **do** set a default, but since play_context fills in its own default before the plugins get a chance to do option processing, they end up with what play_context says.

So I would prefer changing that line in play_context to `'become_exe': self.become_exe or ''` or even just `'become_exe': self.become_exe`.

I'm just not sure how good/bad this would be w.r.t. backwards compatibility. @bcoca @sivel @nitzmahone what do you think?

bcoca commented:
@felixfontein play_context is not currently used for become, that function was left there for backwards compatibility with 3rd party plugins, which i'm fine to force upgrading if they need to use collections.

felixfontein commented:
Ok, fine for me. I've adjusted the PR in community.general so that this function is no longer used (it was only used by the unit tests anyway). So two options:
 1) close this PR;
 2) change this PR so that `resource_from_fqcr` stays (which @sivel wanted), but it won't touch `make_become_cmd` anymore (and the changelog fragments goes too).

What do you all prefer?

gundalow commented:
@felixfontein Have you had any further thoughts or discussions on this?

felixfontein commented:
@gundalow no. I'm fine with both closing this (the original issue is solved anyway) or merging it or a variant (to have a nicer way to access the resource from a FQCR than doing the same "split magic" everywhere).

bcoca commented:
resource_from_fqcr +1

felixfontein commented:
@bcoca @sivel I rebased the PR and removed the original part about become plugins so that it now only adds the `resource_from_fqcr` helper. I also removed the changelog; if you want it back, please tell me :)

felixfontein commented:
/rebuild

felixfontein commented:
@sivel @bcoca @s-hertel thanks for reviewing and merging!
